src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs

```diff
@@ -156,9 +163,96 @@ public async Task<FormulaValue> EvalAsync(string expressionText, CancellationTok
 
             var check = Check(expressionText, (RecordType)parameters.IRContext.ResultType, options);
             check.ThrowOnErrors();
+            try
+            {
+                var newValue = await check.Expression.EvalAsync(parameters, cancel);
+                return newValue;
+            }
+            catch (RuntimeMaxCallDepthException)
+            {
+                return CommonErrors.MaxCallDepth(new IRContext(new Syntax.Span(0, expressionText.Length - 1), FormulaType.BindingError));
+            }
+        }
+
+        /// <summary>
+        /// For private use because we don't want anyone defining a function without binding it.
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="body"></param>
+        /// <param name="returnType"></param>
+        /// <param name="parameters"></param>
+        /// <returns></returns>
+        private UDFLazyBinder DefineFunction(string name, string body, FormulaType returnType, params NamedFormulaType[] parameters)
+        {
+            // $$$ Would be a good helper function 
+            var record = new RecordType();
+            foreach (var p in parameters)
+            {
+                record = record.Add(p);
+            }
+
+            var check = new LazyCheck(this, body, record);
+
+            var func = new UDFTexlFunction(name, returnType, parameters.Select(x => x.Type).ToArray())
+            {
+                _parameterNames = parameters,
+                _check = check,
+            };
+
+            if (_customFuncs.ContainsKey(name))
+            {
+                return new UDFLazyBinder(
+                    new ExpressionError()
+                {
+                    Message = "Function name already defined",
+                    Span = null,
+                    Kind = ErrorKind.Internal, //Todo: change this out for a new kind of ErrorKind error.
+                }, name);

```

Comment:
Should this be an exception instead of runtime error?
Double adding a bug in the host.